docs/proposals/Periodic-Retrieval-of-Certificates-and-Keys-from-AKV.md

```diff
@@ -0,0 +1,121 @@
+# Periodic retrieval of certificates and keys from Azure Key Vault
+
+## Problem/Motivation
+
+When ensuring image integrity and authenticity, you can sign them using either Notation with certificates in Azure Key Vault (AKV) or Cosign with key pairs stored in AKV. To verify these signed images with Ratify in K8s, users typically configure the corresponding certificates or keys using custom resource `KeyManagementProvider`. This allows Ratify to retrieve the correct certificates or keys for verification.
+
+As a security best practice, certificates and keys are rotated regularly. This rotation can occur automatically, resulting in the generation of new versions of the existing certificates or keys in AKV. A certificate may also be revoked due to reasons such as a compromised private key or an invalid certificate identity. In such cases, the certificate is rotated on demand to a new version. Similarly, keys in AKV may be disabled for security reasons, prompting users to rotate them with a new version timely.
+
+Regardless of whether it’s certificate or key rotation, a new version is typically generated for the existing certificate or key. In the current Ratify version (v1.2.0 or prior), when rotation occurs, users must manually apply the `KeyManagementProvider` resource to trigger Ratify to retrieve the latest/specific versions of certificates or keys. This manual step can be cumbersome for users, especially considering automated certificate and key rotation. Users also need to keep the old versions of certificates or keys configured for some periods, as not all images are signed with the latest/specific versions in a short period. In practice, it’s challenging to keep up with these changes manually, and misconfigurations can lead to image verification failures and thus unnecessary service downtime. To address this, automated processes for periodically retrieving all available versions of certificates or keys configured in `KeyManagementProvider` resources would greatly benefit users.
+
+This proposal only covers the case of using AKV as the key management provider. Other providers may have different scenarios that need separate proposals. This proposal does not address the case of verifying images that were signed with Notary Project signatures before the certificate expired, and Ratify handles this case with timestamp support.
```

Comment:
Do we want this proposal to only limit to AKV? Ideally we can extend this to support any KMP and depending on the design we land on, this might not be KMP provider specific.

Reply:
I am not sure we can make it common for all the KMPs, for example, in near feature, there will be KMP for Trusted Signing service, which manages the key and certificates for users, then there is no need for periodic retrieval of keys and certificates. For other KMP similar to AKV, because I did not use them, so I am not sure they use the similar way. For example, AKV will rotate a new version of an existing key or cert, whether it is same for other KMP. Another example is, private key and public key actually they do not expire, however, user can set expire date in AKV, which enhanced the security, if users manage the key by themselves, for example, inline KMP, they cannot do it, and it is less secure.

But I got you point, maybe from design point of view, we can try to design a common framework of periodic retrieval

Reply:
yeah these are good points. I was thinking about it in terms of other cloud key vaults as Binbin mentions. Hashicorp or Amazon in the future.

Reply:
I think further research is warranted. A cursory look at key management solutions in Azure, AWS and Google Cloud shows that rotations (and other changes) generate native platform events which can be integrated with a messaging queue or similar for handling. Ratify could support that workflow by subscribing to the topic, but this proposal is not compatible with this event-based model. Rather, the proposal centers on scheduled polling. This has pros and cons. One pro is that it can be made generic; as long as the KMP is supported by Ratify, and Ratify already handles rate limiting, etc., then it should be possible for Ratify to support rotation scenarios in _any_ KMP, even if suboptimally (as compared to event responses)

Reply:
Thanks @bureado I will add a section for different KMPs. Your comments about event-based model made me think that I need to focus on the scenarios and user experience, that is, what is the best experience for users. Periodic fetch or subscription to events, is more about design and implementation.